cibuildwheel/docker_container.py

```diff
@@ -0,0 +1,189 @@
+import io
+import json
+import os
+import shlex
+import subprocess
+import sys
+import uuid
+from os import PathLike
+from pathlib import Path, PurePath
+from typing import IO, Dict, List, Optional, Sequence, Union
+
+
+class DockerContainer:
+    '''
+    An object that represents a running Docker container.
+
+    Intended for use as a context manager e.g.
+    `with DockerContainer('ubuntu') as docker:`
+
+    A bash shell is running in the remote container. When `call()` is invoked,
+    the command is relayed to the remote shell, and the results are streamed
+    back to cibuildwheel.
+    '''
+    UTILITY_PYTHON = '/opt/python/cp38-cp38/bin/python'
+
+    process: subprocess.Popen
+    bash_stdin: IO[bytes]
+    bash_stdout: IO[bytes]
+
+    def __init__(self, docker_image: str, simulate_32_bit=False):
+        self.docker_image = docker_image
+        self.simulate_32_bit = simulate_32_bit
+
+    def __enter__(self) -> 'DockerContainer':
+        self.name = f'cibuildwheel-{uuid.uuid4()}'
+        shell_args = ['linux32', '/bin/bash'] if self.simulate_32_bit else ['/bin/bash']
+        subprocess.run(
+            [
+                'docker', 'create',
+                '--env', 'CIBUILDWHEEL',
+                '--name', self.name,
+                '-i',
+                '-v', '/:/host',  # ignored on CircleCI
+                self.docker_image,
+                *shell_args
+            ],
+            check=True,
+        )
+        self.process = subprocess.Popen(
+            [
+                'docker', 'start',
+                '--attach', '--interactive',
+                self.name,
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
+
+        assert self.process.stdin and self.process.stdout
+        self.bash_stdin = self.process.stdin
+        self.bash_stdout = self.process.stdout
+
+        # run a noop command to block until the container is responding
+        self.call(['/bin/true'])
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.bash_stdin.close()
+        self.process.terminate()
+        self.process.wait()
+
+        subprocess.run(['docker', 'rm', '--force', '-v', self.name])
+        self.name = None
+
+    def copy_into(self, from_path: Path, to_path: PurePath) -> None:
+        # `docker cp` causes 'no space left on device' error when
+        # a container is running and the host filesystem is
+        # mounted. https://github.com/moby/moby/issues/38995
+        # Use `docker exec` instead.
+
+        if from_path.is_dir():
+            self.call(['mkdir', '-p', to_path])
+            subprocess.run(
+                f'tar cf - . | docker exec -i {self.name} tar -xC {shell_quote(to_path)} -f -',
```

Comment:
Just checking, since you were disappointed by the speed of `docker exec`. Using byte-streams now, it's still not possible to do this with `self.call`?

Reply:
Ooops, I just saw this after the merge. It would be possible, but pretty fiddly to implement. The easiest implementation would involve reading an entire tarball into memory within the host process before streaming it into the container. With a large codebase that might introduce memory issues, and could well be slower. A better implementation would stream between active `tar` processes on either side, but I'm guessing that would involve some pretty gnarly non-blocking I/O code to get it to work right. 

I was curious - here's a profile of the basic test project build - 

```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 19:54:25  Samples:  373
 /_//_/// /_\ / //_// / //_'/ //     Duration: 35.644    CPU time: 0.144
/   _/                      v3.1.3

Program: /Users/joerick/Projects/cibuildwheel/cibuildwheel/__main__.py --platform linux

35.642 <module>  __main__.py:1
└─ 35.599 main  __main__.py:56
   └─ 35.597 build  linux.py:78
      ├─ 25.233 call  docker_container.py:117
      ├─ 3.552 copy_into  docker_container.py:76
      │  └─ 3.541 run  subprocess.py:447
      │        [11 frames hidden]  subprocess
      ├─ 3.184 __enter__  docker_container.py:34
      │  ├─ 2.655 call  docker_container.py:117
      │  └─ 0.520 run  subprocess.py:447
      │        [11 frames hidden]  subprocess
      ├─ 1.854 __exit__  docker_container.py:68
      │  └─ 1.575 wait  subprocess.py:1014
      │        [4 frames hidden]  subprocess
      ├─ 0.682 glob  docker_container.py:106
      │  └─ 0.682 call  docker_container.py:117
      ├─ 0.536 copy_out  docker_container.py:95
      │  └─ 0.536 run  subprocess.py:447
      │        [9 frames hidden]  subprocess
      └─ 0.518 get_environment  docker_container.py:176
         └─ 0.518 call  docker_container.py:117

```

`copy_into` is 10% of the time here. Maybe not worth the hassle.

Reply:
Might not be indeed. It just looks out of place; that's why I'm asked.